### PR TITLE
Refactor PS

### PR DESCRIPTION
**Pull request by UdjinM6**
Some basic refactoring:
- split `CDarksendPool`: server, client and common part.
- `darksend-relay.*` is no longer going to be compiled and linked (same commit, 9cebdc6): it's a legacy implementation of so called "masternode blinding" which never worked properly unfortunately and it was never used in production iirc (at least not for long). Files in src/ are still kept as a reference (and a reminder) however, we might want to revise this later. I tried to keep other changes minimal here.
- split `DoAutomaticDenominating`: moved huge chunks of code into their own functions `CheckAutomaticBackup`, `JoinExistingQueue`, `StartNewQueue`, changes are not as huge as github shows, use `?w=1` to see a bit cleaner diff https://github.com/dashpay/dash/commit/c5c97938254bee99cdbc30d458d52c87e659827e?w=1
- `CMasternode*` -> `masternode_info_t` for most parts in PS client.

Slightly tested, works correctly both as a client and as a MN for me.

### CODE REVIEW

**Comment by tgflynn** [comment]
:+1: 

Generally looks really good.

I would suggest moving non-const global variables like these:

`
extern int nPrivateSendRounds;
extern int nPrivateSendAmount;
extern int nLiquidityProvider;
extern bool fEnablePrivateSend;
extern bool fPrivateSendMultiSession;
`

into the top-level objects so we can cut down on the number of global variables.

These should also be protected for MT access but I think the best way to do that is probably to use std::atomic from C++-11, so that could be done in a later PR after the C++-11 build has been enabled.

**Comment by UdjinM6** [comment]
Good point! 👍

**Comment by taw00** [comment]
Really glad to see a focus on architecture with the code. I.e., Making it better and more maintainable. Kudos, guys.

File: src/privatesend-client.cpp [hunk 1]
```diff
@@ -1343,9 +1344,9 @@ bool CPrivateSendClient::CreateDenominated(const CompactTallyItem& tallyItem, bo
 
 void CPrivateSendClient::RelayIn(const CDarkSendEntry& entry)
 {
-    if(!pSubmittedToMasternode) return;
+    if(!infoMixingMasternode.fInfoValid) return;
 
-    CNode* pnode = FindNode(pSubmittedToMasternode->addr);
+    CNode* pnode = FindNode(infoMixingMasternode.addr);
```
Comment (tgflynn) [inline src/privatesend-client.cpp]: I think this should lock cs_vNodes (like other calls to FindNode in this file do).

File: src/privatesend-client.cpp [hunk 2]
```diff
@@ -937,11 +938,11 @@ bool CPrivateSendClient::StartNewQueue(CAmount nValueMin, CAmount nBalanceNeedsA
             }
         }
 
-        LogPrintf("CPrivateSendClient::StartNewQueue -- attempt %d connection to Masternode %s\n", nTries, pmn->addr.ToString());
-        CNode* pnode = (pnodeFound && pnodeFound->fMasternode) ? pnodeFound : ConnectNode((CAddress)pmn->addr, NULL, true);
+        LogPrintf("CPrivateSendClient::StartNewQueue -- attempt %d connection to Masternode %s\n", nTries, infoMn.addr.ToString());
+        CNode* pnode = (pnodeFound && pnodeFound->fMasternode) ? pnodeFound : ConnectNode((CAddress)infoMn.addr, NULL, true);
```
Comment (tgflynn) [inline src/privatesend-client.cpp]: If ConnectNode gets called here how do we know pnode gets released ?

File: src/privatesend-client.cpp [hunk 3]
```diff
@@ -94,7 +94,7 @@ void CPrivateSendClient::ProcessMessage(CNode* pfrom, std::string& strCommand, C
             pmn->fAllowMixingTx = true;
 
             LogPrint("privatesend", "DSQUEUE -- new PrivateSend queue (%s) from masternode %s\n", dsq.ToString(), pmn->addr.ToString());
-            if(pSubmittedToMasternode && pSubmittedToMasternode->vin.prevout == dsq.vin.prevout) {
+            if(infoMixingMasternode.vin.prevout == dsq.vin.prevout) {
```
Comment (tgflynn) [inline src/privatesend-client.cpp]: The equivalent logic would be to first check infoMixingMasternode.fInfoValid.

File: src/masternodeman.cpp [hunk 4]
```diff
@@ -766,7 +766,7 @@ void CMasternodeMan::ProcessMasternodeConnections()
     LOCK(cs_vNodes);
     BOOST_FOREACH(CNode* pnode, vNodes) {
         if(pnode->fMasternode) {
-            if(privateSendClient.pSubmittedToMasternode != NULL && pnode->addr == privateSendClient.pSubmittedToMasternode->addr) continue;
+            if(pnode->addr == privateSendClient.infoMixingMasternode.addr) continue;
```
Comment (tgflynn) [inline src/masternodeman.cpp]: The equivalent logic would check fInfoValid.

**Comment by tgflynn** [comment]
Sorry, change above comment to "should use mnodeman.GetMasternodeInfo()".

**Comment by UdjinM6** [comment]
Not sure why github thinks comments are outdated even before I pushed any updates, so to avoid notification spam I'll just answer here.
privatesend-client.cpp:
L49: agree
L97: agree
L941: it's already handled in old code, not sure why you are asking here. Or maybe I do not quite understand the question.
L1349: agree
masternodeman.cpp:
L769: agree

**Comment by tgflynn** [comment]
RE: Line 941.  I guess I don't quite understand the logic here.

If the node is found in vNodes then this block of code sends DSACCEPT and releases the node.

However if the node is not already in vNodes it connects to it and sends the message but does not then release the node.

Perhaps not directly related to the PR but I'm wondering if this isn't a bug.

**Comment by UdjinM6** [comment]
@tgflynn Ah, I see. So the logic here is this: if this node is already connected, we add a reference to it regardless if it's a MN or not. If it's not connected or it's not a MN we connect to it/update it `fMasternode` flag and add a reference inside of `ConnectNode`. Connections to MNs are handled in `ProcessMasternodeConnections` and reference is released later in `ThreadSocketHandler`.

**Pull request merged**